src/sagas/networkSettings.js

```diff
@@ -0,0 +1,233 @@
+  // In practice they will never be equals.
+
+  const backupUrl = {
+    nodeUrl: config.getServerUrl(),
+    explorerServiceUrl: config.getExplorerServiceBaseUrl(),
+    walletServiceUrl: config.getWalletServiceBaseUrl(),
+    walletServiceWsUrl: config.getWalletServiceBaseWsUrl(),
+  };
+
+  config.setServerUrl(nodeUrl);
+  config.setExplorerServiceBaseUrl(explorerServiceUrl);
+
+  // - walletServiceUrl has precedence
+  // - nodeUrl as fallback
+  let network;
+  if (walletServiceUrl) {
+    config.setWalletServiceBaseUrl(walletServiceUrl);
+    config.setWalletServiceBaseWsUrl(walletServiceWsUrl);
+
+    try {
+      network = yield call(getWalletServiceNetwork);
+    } catch (err) {
+      rollbackConfigUrls(backupUrl);
+      yield put(networkSettingsUpdateFailure());
+      return;
+    }
+  }
+
+  if (!network) {
+    try {
+      network = yield call(getFullnodeNetwork);
+    } catch (err) {
+      rollbackConfigUrls(backupUrl);
+      yield put(networkSettingsUpdateFailure());
+      return;
+    }
+  }
+
+  // Fail after try get network from fullnode
+  if (!network) {
+    yield put(networkSettingsUpdateFailure());
+    return;
+  }
+
+  let stage;
+  if (network === NETWORK_TESTNET) {
+    stage = STAGE_TESTNET;
+  } else if (network === NETWORK) {
+    stage = STAGE;
+  } else {
+    stage = STAGE_DEV_PRIVNET;
+  }
+
+  const customNetwork = {
+    stage,
+    network,
+    nodeUrl,
+    explorerUrl,
+    explorerServiceUrl,
+    walletServiceUrl,
+    walletServiceWsUrl,
+  };
+
+  yield put(networkSettingsUpdateSuccess(customNetwork));
+}
+
+/**
+ * Rollback the URLs configured in the wallet by using the backupUrl object.
+ * @param {{
+ *   nodeUrl: string;
+ *   explorerServiceUrl: string;
+ *   walletServiceUrl: string;
+ *   walletServiceWsUrl: string;
+ * }} backupUrl An object containing the previous configuration for wallet URLs.
+ */
+function rollbackConfigUrls(backupUrl) {
+  config.setServerUrl(backupUrl.nodeUrl);
+  config.setExplorerServiceBaseUrl(backupUrl.explorerServiceUrl);
+  config.setWalletServiceBaseUrl(backupUrl.walletServiceUrl);
+  config.setWalletServiceBaseWsUrl(backupUrl.walletServiceWsUrl);
+}
+
+/**
+ * Check if the value is an invalid URL.
+ *
+ * @param {string} value to be checked.
+ * @returns {boolean} true if invalid, false otherwise.
+ */
+function invalidUrl(value) {
+  const allowedProtocols = ['http:', 'https:', 'ws:', 'wss:'];
+  try {
+    const tryUrl = new URL(value);
+    if (isEmpty(tryUrl.protocol) || !(allowedProtocols.includes(tryUrl.protocol))) {
+      return true;
+    }
+
+    if (isEmpty(tryUrl.host)) {
+      return true;
+    }
+  } catch {
+    return true;
+  }
+
+  return false;
+}
+
+/**
+ * Saves the network custom settings in the application storage.
+ * @param {object} action contains the payload with the network settings
+ * already saved in the redux store.
+ */
+export function* persistNetworkSettings(action) {
```

Comment:
Why don't you update the redux state as well here? We would have a different information in the redux and in the storage?

Reply:
No. I separated both to distinguish the input treatment from the storing process. Between the update handover to the persist function, the redux stores the value in the state. That's when we have the opportunity to change the interface UI almost immediately. But of course, this can be changed, it is just a preference to segregate the processes.